### PR TITLE
avoid string conversion while doing timestamp parsing whenever possible

### PR DESCRIPTION
**Pull request by nishantmonu51**
avoid string conversion while doing timestamp parsing whenever possible.

### CODE REVIEW

**Comment by xvrl** [comment]
it might make sense to create a separate timestamp parser for Long values instead of lumping it together in the existing one. That'll preserve backwards compatibility and make it clearer I hope.

**Comment by nishantmonu51** [comment]
have made changes to the api to preserve backwards compatible. 
Please have a look again.

File: src/main/java/com/metamx/common/parsers/TimestampParser.java [hunk 1]
```diff
@@ -28,76 +28,98 @@
...
         @Override
-        public DateTime apply(String input)
+        public DateTime apply(Object input)
         {
-          Preconditions.checkArgument(input != null && !input.isEmpty(), "null timestamp");
-          return new DateTime(Long.parseLong(ParserUtils.stripQuotes(input)) * 1000);
+          Preconditions.checkArgument(input != null, "null timestamp");
+          String str = input.toString();
+          Preconditions.checkArgument(!str.isEmpty(), "empty timestamp");
+          return new DateTime(Long.parseLong(ParserUtils.stripQuotes(str)) * 1000);
```
Comment (himanshug) [inline src/main/java/com/metamx/common/parsers/TimestampParser.java]: why are we not doing `instanceof Number` check in this case?

File: src/main/java/com/metamx/common/parsers/TimestampParser.java [hunk 2]
```diff
@@ -28,76 +28,98 @@
...
         @Override
-        public DateTime apply(String input)
+        public DateTime apply(Object input)
         {
-          Preconditions.checkArgument(input != null && !input.isEmpty(), "null timestamp");
-          Double ts = Double.parseDouble(ParserUtils.stripQuotes(input));
+          Preconditions.checkArgument(input != null , "null timestamp");
+          String str = input.toString();
+          Preconditions.checkArgument(!str.isEmpty(), "empty timestamp");
+          Double ts = Double.parseDouble(ParserUtils.stripQuotes(str));
```
Comment (himanshug) [inline src/main/java/com/metamx/common/parsers/TimestampParser.java]: why are we not doing instanceof Number check in this case?

File: src/main/java/com/metamx/common/parsers/TimestampParser.java [hunk 3]
```diff
@@ -28,76 +28,107 @@
 {
   private static final Logger log = new Logger(TimestampParser.class);
 
-  public static Function<String, DateTime> createTimestampParser(
+  public static Function<Object, DateTime> createTimestampParserGeneric(
```
Comment (xvrl) [inline src/main/java/com/metamx/common/parsers/TimestampParser.java]: I wonder if we should leave the existing one as is, and create new one that is specific to Long values. That would probably cleaner, remove a lot of branching in this method, and also leave the signature more specific as to which types we expect.

Comment (nishantmonu51) [inline src/main/java/com/metamx/common/parsers/TimestampParser.java]: i tried doing that but its not same for all the cases, nano and ruby have different logic since the Number cannot be directly parsed to timestamp.

**Comment by nishantmonu51** [comment]
@xvrl @himanshug have handled the review comments, please have a look.

File: src/main/java/com/metamx/common/parsers/TimestampParser.java [hunk 4]
```diff
@@ -93,9 +93,11 @@ public DateTime apply(String input)
         }
       };
     } else if (format.equalsIgnoreCase("nano")) {
-      return new Function<String, DateTime>() {
+      return new Function<String, DateTime>()
```
Comment (himanshug) [inline src/main/java/com/metamx/common/parsers/TimestampParser.java]: it appears that 4 "else if" for "posix", "ruby", "nano" and "mills" could be replaced by single else if like
```
..
else if (format.equalsIgnoreCase("posix"))  ||
           format.equalsIgnoreCase("ruby"))  ||
           format.equalsIgnoreCase("millis"))  ||
           format.equalsIgnoreCase("nano")) ) {
    return new Function<String, DateTime>() {
        private Function<Number, DateTime> func = createNumericTimestampParser(format);
        public DateTime apply(String input) {
            return func.apply(Long.parseLong(ParserUtils.stripQuotes(input)));
        }
    }
} 
else ...
```

Comment (nishantmonu51) [inline src/main/java/com/metamx/common/parsers/TimestampParser.java]: done. refactored.

**Comment by himanshug** [comment]
LGTM

File: src/main/java/com/metamx/common/parsers/TimestampParser.java [hunk 5]
```diff
@@ -60,46 +60,28 @@ public DateTime apply(String input)
...
       {
         @Override
         public DateTime apply(String input)
         {
           Preconditions.checkArgument(input != null && !input.isEmpty(), "null timestamp");
-          return new DateTime(Long.parseLong(ParserUtils.stripQuotes(input)) * 1000);
+          return numericFun.apply(Long.parseLong(ParserUtils.stripQuotes(input)));
         }
       };
     } else if (format.equalsIgnoreCase("ruby")) {
```
Comment (xvrl) [inline src/main/java/com/metamx/common/parsers/TimestampParser.java]: any reason ruby is still separate from the other numeric ones?

Comment (drcrallen) [inline src/main/java/com/metamx/common/parsers/TimestampParser.java]: Does that change fit within the scope of this PR, or would that be better for a different PR?

Comment (nishantmonu51) [inline src/main/java/com/metamx/common/parsers/TimestampParser.java]: ruby supports Double but other formats only long.

Comment (xvrl) [inline src/main/java/com/metamx/common/parsers/TimestampParser.java]: ok, I see it now. I'm surprised it's actually working given the fact that createNumericTimestampParser uses input.longValue() for ruby format as far as I can see.

Comment (nishantmonu51) [inline src/main/java/com/metamx/common/parsers/TimestampParser.java]: the ruby parser igores millis. 
prev impl - 
 Preconditions.checkArgument(input != null && !input.isEmpty(), "null timestamp");
          Double ts = Double.parseDouble(ParserUtils.stripQuotes(input));
          Long jts = ts.longValue() * 1000; // ignoring milli secs // done in numeric parser
          return new DateTime(jts);     /// in numeric parser now. 

Refactoring it to be more clear now.

Comment (nishantmonu51) [inline src/main/java/com/metamx/common/parsers/TimestampParser.java]: TimestampParserTest.testRuby verifies ruby implementation. 
Added comments in code.
